File: backend/app/domain/depot_assignment.py
from typing import List, Optional, NamedTuple
from dataclasses import dataclass

@dataclass
class DepotInfo:
    id: str
    name: str
    zone_code: str
    daily_capacity: int
    current_order_count: int
    pincodes: List[str]
    is_active: bool = True

class DepotAssignmentResult(NamedTuple):
    assigned_depot_id: Optional[str]
    assigned_depot_name: Optional[str]
    assignment_type: str  # 'primary', 'fallback', or 'unassigned'
    reason: str

def assign_depot_for_order(
    customer_pincode: str,
    available_depots: List[DepotInfo]
) -> DepotAssignmentResult:
    """
    Pure deterministic multi-depot routing engine.
    Matches customer pincode to active depots with daily capacity.
    Falls back to secondary depot if primary is full.
    Flags order as unassigned if no eligible depot is found.
    """
    active_depots = [d for d in available_depots if d.is_active]

    # Primary depot match by pincode
    primary_depots = [d for d in active_depots if customer_pincode in d.pincodes]

    for depot in primary_depots:
        if depot.current_order_count < depot.daily_capacity:
            return DepotAssignmentResult(
                assigned_depot_id=depot.id,
                assigned_depot_name=depot.name,
                assignment_type="primary",
                reason=f"Assigned to primary depot {depot.name} for pincode {customer_pincode}."
            )

    # Fallback: check other active depots with available capacity
    fallback_depots = [d for d in active_depots if d not in primary_depots]
    for depot in fallback_depots:
        if depot.current_order_count < depot.daily_capacity:
            return DepotAssignmentResult(
                assigned_depot_id=depot.id,
                assigned_depot_name=depot.name,
                assignment_type="fallback",
                reason=f"Primary depot at capacity; reassigned to secondary depot {depot.name}."
            )

    # Unassigned Exception
    return DepotAssignmentResult(
        assigned_depot_id=None,
        assigned_depot_name=None,
        assignment_type="unassigned",
        reason=f"No active depot has capacity for pincode {customer_pincode}. Routed to Admin Exceptions Queue."
    )

File: src/domain/depotAssignment.ts
import { Depot, DepotAssignment, Address, DeliverySlot } from '../types';

export type DepotAssignmentResult = {
  success: boolean;
  assignment: DepotAssignment;
  selectedDepot?: Depot;
  reason: string;
};

/**
 * Deterministic multi-depot assignment algorithm.
 * 
 * Rules:
 * 1. Filter active depots with slot support & available capacity.
 * 2. Check for depots serving the customer's pincode / zone.
 * 3. If primary zone depots exist and have capacity, pick the best primary depot.
 * 4. If primary zone depots are full/unavailable, fallback to eligible secondary depots with capacity.
 * 5. If no depot qualifies, mark assignment as failed -> creates exception.
 */
export function assignDepotForOrder(
  address: Address,
  deliverySlot: DeliverySlot,
  depots: Depot[]
): DepotAssignmentResult {
  const activeDepots = depots.filter(d => d.isActive);
  
  if (activeDepots.length === 0) {
    return {
      success: false,
      assignment: {
        primaryDepotId: '',
        assignmentType: 'fallback',
        assignmentReason: 'No active depots in the system',
      },
      reason: 'No active depots in the system',
    };
  }

  // Filter depots with available capacity & slot support
  const availableDepots = activeDepots.filter(d => 
    d.supportedSlots.includes(deliverySlot) &&
    d.currentOrderCount < d.dailyOrderCapacity
  );

  if (availableDepots.length === 0) {
    const primaryDepot = activeDepots[0];
    return {
      success: false,
      assignment: {
        primaryDepotId: primaryDepot.id,
        assignmentType: 'fallback',
        assignmentReason: 'All system depots are at full order capacity for this slot',
      },
      reason: 'All system depots are at full order capacity for this slot',
    };
  }

  // Check if any available depot serves the pincode/zone directly
  const primaryZoneDepots = availableDepots.filter(d => 
    d.serviceZones.includes(address.pincode) || 
    d.serviceZones.some(z => address.street.toLowerCase().includes(z.toLowerCase()) || address.city.toLowerCase().includes(z.toLowerCase()))
  );

  const isPrimaryMatch = primaryZoneDepots.length > 0;
  const candidateDepots = isPrimaryMatch ? primaryZoneDepots : availableDepots;

  // Pick candidate depot with lowest utilization ratio
  const bestDepot = candidateDepots.reduce((prev, curr) => {
    const prevRatio = prev.currentOrderCount / prev.dailyOrderCapacity;
    const currRatio = curr.currentOrderCount / curr.dailyOrderCapacity;
    return currRatio < prevRatio ? curr : prev;
  });

  return {
    success: true,
    selectedDepot: bestDepot,
    assignment: {
      primaryDepotId: isPrimaryMatch ? bestDepot.id : (activeDepots.find(d => d.serviceZones.includes(address.pincode))?.id || bestDepot.id),
      assignedDepotId: bestDepot.id,
      assignmentType: isPrimaryMatch ? 'automatic' : 'fallback',
      assignmentReason: isPrimaryMatch ? 'Optimal zone match & capacity' : 'Secondary fallback depot with capacity',
      assignedAt: new Date().toISOString(),
    },
    reason: `Assigned to ${bestDepot.name}`,
  };
}

/**
 * Manual Admin Override for Depot Reassignment.
 */
export function overrideDepotAssignment(
  currentAssignment: DepotAssignment,
  newDepotId: string,
  reason: string,
  adminId: string
): DepotAssignment {
  if (!reason || reason.trim().length === 0) {
    throw new Error('Mandatory override reason is required');
  }

  return {
    ...currentAssignment,
    assignedDepotId: newDepotId,
    assignmentType: 'manual',
    assignmentReason: reason.trim(),
    assignedAt: new Date().toISOString(),
    assignedBy: adminId,
  };
}
